### PR TITLE
Added eval metrics and fixed bugs

### PR DESCRIPTION
**Pull request by baitsguy**
1. Added document and page retrieval metrics
2. Added rouge metrics for generated answers
3. Added a "page_numbers" field when merging elements to collect list of pages used to build an element
4. Fixed bugs in field names

### CODE REVIEW

File: sycamore/evaluation/pipeline.py [hunk 1]
```diff
@@ -1,116 +1,143 @@
...
+        if "llm" in self._os_config:
+            query["params"] = {"search_pipeline": self._os_config["search_pipeline"]}
+            query["query"]["ext"] = {
+                "generative_qa_parameters": {
+                    "llm_question": doc["question"],
+                    "context_size": self._os_config["context_window"],
+                    "llm_model": self._os_config["llm"],
                 },
-                "size": self._os_config["size"],
+                "rerank": {"query_context": {"query_text": doc["question"]}},
```
Comment (HenryL27) [inline sycamore/evaluation/pipeline.py:87]: Oh we're assuming rerank then?

File: sycamore/evaluation/pipeline.py [hunk 2]
```diff
@@ -1,116 +1,143 @@
...
+        if "filters" in doc:
+            query["query"] = self._add_filter(query["query"], doc["filters"])
         return query
 
-    def _evaluate_queries(self, query_result: Document) -> Document:
+    def _process_queries(self, query_result: Document) -> Document:
+        logger.debug(("Query result: " + str(query_result.keys())))
         assert query_result.type == "OpenSearchQueryResult"
         result = EvaluationDataPoint(query_result)
+        result.generated_source_documents = [Element(hit) for hit in query_result["hits"]]
```
Comment (HenryL27) [inline sycamore/evaluation/pipeline.py:97]: think this need to be `... hit in query_results['hits']['hits']`?

File: sycamore/evaluation/pipeline.py [hunk 3]
```diff
@@ -1,116 +1,143 @@
...
+        for metric in self._metrics:
+            metric_data[metric.metric_name()] = {}
+            summary.metrics[metric.metric_name()] = {}
+
+        for doc in query_level_metrics.take_all():
+            for metric_key, metric_value in metric_data.items():
+                for sub_metric_key, sub_metric_value in doc["metrics"].get(metric_key, {}).items():
+                    metric_data[metric_key][sub_metric_key] = metric_data[metric_key].get(sub_metric_key, []) + [
+                        sub_metric_value
+                    ]
```
Comment (HenryL27) [inline sycamore/evaluation/pipeline.py:123]: there's gotta be a fancy ray mapreduce-y/spark-y way to do this

**Review by HenryL27 — COMMENTED** [review]
overall looks good. My main worry is about how the specific things we've done for customers generalize to library-level code

File: sycamore/evaluation/data.py [hunk 4]
```diff
@@ -34,14 +34,14 @@ def ground_truth_answer(self, value: str) -> None:
...
-        """Source document url."""
-        return self.data.get("ground_truth_document_url")
+    def filters(self) -> Optional[dict[str, str]]:
+        """Filters to use for querying"""
+        return self.data.get("filters")
 
-    @ground_truth_document_url.setter
-    def ground_truth_document_url(self, value: str) -> None:
-        """Set the document url."""
-        self.data["ground_truth_document_url"] = value
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply sycamore/evaluation/data.py:44]: it was confusing to have this + ground_truth_search_documents so just kept 1

File: sycamore/evaluation/data.py [hunk 5]
```diff
@@ -16,7 +16,7 @@ def __init__(
     @property
     def question(self) -> Optional[str]:
         """Natural language question."""
-        return self.data.get("query")
+        return self.data.get("question")
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply sycamore/evaluation/data.py:19]: This is the question. The OSrequest object stores the os query

File: sycamore/evaluation/data.py [hunk 6]
```diff
@@ -83,6 +83,16 @@ def metrics(self, value: dict[str, Any]) -> None:
...
 
+    @property
+    def additional_info(self) -> Optional[dict[str, Any]]:
+        """Additional information for specialized datapoints."""
+        return self.data.get("additional_info")
+
+    @additional_info.setter
+    def additional_info(self, value: Optional[dict[str, Any]]) -> None:
+        """Set the additional information for specialized datapoints."""
+        self.data["additional_info"] = value
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply sycamore/evaluation/data.py:94]: disable_rewriting/ndd was one

File: sycamore/evaluation/metrics.py [hunk 7]
```diff
@@ -12,5 +18,76 @@ def metric_name(self) -> str:
...
+                    in_page_numbers = self._match_any_page and ground_truth_page_number in document.properties.get(
+                        "page_numbers", {}
+                    )
+                    if (
+                        in_page_numbers
+                        or document.properties.get("page_number", -1) == ground_truth_document.properties["page_number"]
+                    ):
+                        result["correct_page_position"] = result.get("correct_page_position", i + 1)
+                        page_mrr_sum += 1.0 / (i + 1.0)
+                        correct_page_count += 1
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply sycamore/evaluation/metrics.py:63]: Seems pretty generic for any page number based dataset? Anything specific you want to change?

File: sycamore/evaluation/metrics.py [hunk 8]
```diff
@@ -12,5 +18,76 @@ def metric_name(self) -> str:
...
+
+        result["doc_recall"] = 1.0 if correct_doc_count > 0 else 0.0
+        result["page_recall"] = 1.0 if correct_page_count > 0 else 0.0
+        result["doc_mrr"] = min(1.0, doc_mrr_sum / len(datapoint.ground_truth_source_documents))
+        result["page_mrr"] = min(1.0, page_mrr_sum / len(datapoint.ground_truth_source_documents))
+        return result
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply sycamore/evaluation/metrics.py:69]: I thought about expanding to it as this set grows. Right now these are lightweight and reasonable defaults to always publish

File: sycamore/evaluation/metrics.py [hunk 9]
```diff
@@ -12,5 +18,76 @@ def metric_name(self) -> str:
...
+
+
+class GeneratedAnswerMetrics(EvaluationMetric):
+    def __init__(self, rouge_metrics=None) -> None:
+        super().__init__()
+        if rouge_metrics is None:
+            rouge_metrics = ["rouge-1", "rouge-2", "rouge-l"]
+        self._rouge_evaluator = rouge.Rouge(metrics=rouge_metrics)
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply sycamore/evaluation/metrics.py:77]: Same as above, I thought about expanding to it as this set grows. Right now these are lightweight and reasonable defaults to always publish

File: sycamore/evaluation/pipeline.py [hunk 10]
```diff
@@ -1,116 +1,143 @@
...
+            "query": {
+                "hybrid": {
+                    "queries": [
+                        {"match": {"text_representation": doc["question"]}},
+                        {
+                            "neural": {
+                                "embedding": {
+                                    "query_text": doc["question"],
+                                    "model_id": self._os_config["embedding_model_id"],
+                                    "k": self._os_config["neural_search_k"],
```
[the remark this answers is not in the document]
Comment (baitsguy) [reply sycamore/evaluation/pipeline.py:68]: can do

File: sycamore/evaluation/pipeline.py (hunk 1 above)
Comment (baitsguy) [reply sycamore/evaluation/pipeline.py:87]: Yeah this is basically our "default" os query. I'm sure we'll have cases where we'll want to customize this

File: sycamore/evaluation/pipeline.py (hunk 2 above)
Comment (baitsguy) [reply sycamore/evaluation/pipeline.py:97]: The term might be a little confusing, this is the internal query_result object of type (OpenSearchQueryResult). That has a hit property that is populated when parsing the raw os result in OpenSearchQueryExecutor

File: sycamore/evaluation/pipeline.py (hunk 3 above)
Comment (baitsguy) [reply sycamore/evaluation/pipeline.py:123]: Yeah I have some references from Ben how to do a group/aggregate op but it was a little complex/overkill for this case. Will evaluate later

File: sycamore/evaluation/metrics.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (HenryL27) [reply sycamore/evaluation/metrics.py:63]: not really. I'm just worried that someone will plug a dataset into this and get a whole bunch of key errors or something. I think I'm just paranoid tho

**Review by HenryL27 — APPROVED** [review]

File: sycamore/evaluation/metrics.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (baitsguy) [reply sycamore/evaluation/metrics.py:63]: Yeah this assumes "page_number" is present in the input/output

**Pull request merged**